app/src/main/java/com/alphawallet/app/viewmodel/TokenFunctionViewModel.java

```diff
@@ -718,75 +742,79 @@ private Intent completeTransferIntent(Context ctx, Token token, List<BigInteger>
-            JSONObject result = new JSONObject(asset);
-            if (result.has("collection") && result.has("traits"))
-            {
-                computeRarity(getCountFromJson(result), getTraitsFromJson(result));
-            }
-            else
-            {
-                useTraitsFromNFTAsset(token, tokenId);
-            }
-        }
-        catch (JSONException e)
-        {
-            Timber.e(e);
-        }
+        Timber.e(t);
     }
 
-    private void useTraitsFromNFTAsset(Token token, BigInteger tokenId)
+    private void onAssetMetadata(NFTAsset asset)
     {
-        NFTAsset nftAsset = token.getAssetForToken(tokenId);
-        if (nftAsset != null)
-        {
-            attrs.postValue(nftAsset);
-        }
+        nftAsset.postValue(asset);
     }
 
-    private ArrayList<Trait> getTraitsFromJson(JSONObject result) throws JSONException
+    public void getAsset(Token token, BigInteger tokenId)
     {
-        return new Gson().fromJson(result.getString("traits"),
-                new TypeToken<ArrayList<Trait>>()
-                {
-                }.getType());
+        openseaDisposable = openseaService.getAsset(token, tokenId)
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(result -> onAsset(result, token, tokenId), this::onAssetError);
     }
 
-    private long getCountFromJson(JSONObject result) throws JSONException
+    private void onAssetError(Throwable throwable)
     {
-        JSONObject collectionJson = result.getJSONObject("collection");
-        Stats stats = new Gson().fromJson(collectionJson.getString("stats"), Stats.class);
-        return stats.getCount();
+        Timber.d(throwable);
     }
 
-    private void computeRarity(long count, List<Trait> traitList) throws JSONException
+    private void onAsset(String result, Token token, BigInteger tokenId)
     {
-        for (Trait trait : traitList)
+        boolean loadedFromApi = false;
+        if (JsonUtils.isValidAsset(result))
         {
-            if (trait.getTraitCount() == 1)
+            try
+            {
+                JSONObject assetJson = new JSONObject(result);
+                OpenSeaAsset osAsset = new Gson().fromJson(assetJson.toString(), OpenSeaAsset.class);
+                if (osAsset != null)
+                {
+                    openSeaAsset.postValue(osAsset);
+                    loadedFromApi = true;
+                }
+            }
+            catch (JSONException e)
             {
-                trait.setUnique(true);
+                Timber.e(e);
+                Timber.d("Error fetching from OpenSea: %s", result);
+            }
+            catch (Exception e)
+            {
+                Timber.e(e);
             }
-
-            trait.updateTraitRarity(count);
         }
 
-        traits.postValue(traitList);
+        if (!loadedFromApi)
+        {
+            NFTAsset a = token.getAssetForToken(tokenId);
```

Comment:
I think this is the wrong way round,

Since we are re-loading metadata, if the API call fails we should then fall back to the direct contract call, and if that fails then re-send the old data from the token onAssetMetadataError